C++: Generate int-to-bool conversions in C code

MathiasVP opened the pull request:
In C++ code we have int-to-bool conversions on `myInt` in examples like:
```cpp
void test(int myInt) {
  if(myInt) {
   // ...
  }
}
```
That is, it's not `myInt` that is branched on, but rather the result of `myInt != 0`. In the IR, this manifests as the IR looking like:
```cpp
r1_5(glval<int>) = VariableAddress[myInt]     :
m1_6(int)        = InitializeParameter[myInt] : &:r1_5
r2_1(glval<int>) = VariableAddress[myInt]     :
r2_2(int)        = Load[myInt]                : &:r2_1, m1_6
r2_3(int)        = Constant[0]                :
r2_4(bool)       = CompareNE                  : r2_2, r2_3
v2_5(void)       = ConditionalBranch          : r2_4
```
(notice the `BranchNE` instruction).

This information is important in guard conditions since branching on `myInt` doesn't tell us that `myInt == 1`, but rather that `(myInt != 0) == 1`.

This is all well and good in C++. However, in C, there are no int-to-bool conversions. So when compiling the same code with a C compiler the IR looks like:
```cpp
r1_5(glval<int>) = VariableAddress[myInt]     :
m1_6(int)        = InitializeParameter[myInt] : &:r1_5
r2_1(glval<int>) = VariableAddress[myInt]     :
r2_2(int)        = Load[myInt]                : &:r2_1, m1_6
v2_5(void)       = ConditionalBranch          : r2_2
```
(notice that we're now branching on the value of `myInt`)

Historically, we've worked around this problem in IRGuards specifically (see https://github.com/github/codeql/pull/16364 and https://github.com/github/codeql/pull/16533), but upon facing a related problem somewhere else I now feel like we should simply make the IR equivalent in C and C++.

So this PR inserts IR equivalent to what would have been generated for an int-to-bool conversion if it had been present in the database in the places I'm aware of this being a problem. It then removes the workarounds in the guards library that were added to handle these.

Commit-by-commit review recommended (especially for the IR construction related commits).

jketema commented:
It looks like the changes in https://github.com/github/codeql/pull/18490/commits/65771614ef89e6664649ace7d7b5ac5e8fe91d18 only affect syntax-zoo, and not the IR tests. I think it would make sense to add an IR test, so we can easily see that the generated IR is actually correct.

jketema commented on cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:
```diff
@@ -233,6 +233,125 @@ private class GuardConditionFromBinaryLogicalOperator extends GuardConditionImpl
+  }
+
+  pragma[inline]
+  override predicate comparesLt(Expr left, Expr right, int k, boolean isLessThan, boolean testIsTrue) {
+    exists(Instruction li, Instruction ri |
+      li.getUnconvertedResultExpression() = left and
+      ri.getUnconvertedResultExpression() = right and
+      ir.comparesLt(li.getAUse(), ri.getAUse(), k, isLessThan, testIsTrue.booleanNot())
+    )
+  }
```
I need some more explanation here. Does this predicate (and the ones below) every produce anything sensible given that `ir` is a `CompareEQ`?

MathiasVP replied:
Yeah, I doubt this ever produces anything, really. I tried to quick-eval this predicate on a couple of local DBs, but nothing came up.

Would you prefer just doing `none` here instead?

MathiasVP commented:
> It looks like the changes in [6577161](https://github.com/github/codeql/commit/65771614ef89e6664649ace7d7b5ac5e8fe91d18) only affect syntax-zoo, and not the IR tests. I think it would make sense to add an IR test, so we can easily see that the generated IR is actually correct.

Good point. I've added this in 2076c1c51fd230ff3136db80e4a9252305ba9ec8. I had to add another file since the existing `ir.c` is marked as `--microsoft` which doesn't support the binary conditional operator.

MathiasVP commented:
Thanks for the chat, @jketema! Some comments on the commits I just pushed:
- I've added the reference example we saw on MRVA in 54faba21ba36f3d805ad670ca1b7bc82ac8d785f.
- I removed the type test in d0bd6eb2eb73e9f585ef05423bbedd0d25faa33f. This gives some new good results in the tests for C++ as well (where the operand of the not has a `BoolType` which meant it was previously not part of this class).

After I removed the type test requirement I reran MRVA and looked for cases where we had results for unary equality (we had _lots_), binary equality, unary "less than"s, and binary "less than"s. All of these cases had quite a few results, and I've added an example of each case in d5b31eb98cb136c20f5662a6ff3f5e6b946daa0b.

So, unlike what I expressed in our call earlier, I'm not actually sure we should replace the overrides with `none()` since doing so breaks all the tests added in d5b31eb98cb136c20f5662a6ff3f5e6b946daa0b.

What do you think?

jketema commented on cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:
````diff
@@ -233,6 +233,124 @@ private class GuardConditionFromBinaryLogicalOperator extends GuardConditionImpl
+  IRGuardCondition ir;
+
+  GuardConditionFromNotExpr() {
+    // When `!` is applied to an integer (such as `x`) the generated IR looks
+    // like:
+    // ```
+    // r1(glval<int>) = VariableAddress[myInt] :
+    // r2(int)        = Load[x]                : &:r1, m1_6
+    // r3(int)        = Constant[0]            :
+    // r4(bool)       = CompareEQ              : r2, r3
+    // ```
+    // And so the `IRGuardCondition` for an expression such as `if(!x)` is the
+    // `CompareEQ` instruction. However, users often expect the `x` to also
+    // be a guard condition. But from the perspective of the IR the `x` is just
+    // the left-hand side of a comparison against 0 so it's not included as a
+    // normal `IRGuardCondition`. So to align with user expectations we make
+    // that `x` a `GuardCondition`.
````
This is no longer up-to-date, I think, as you dropped the non-boolean restriction.

MathiasVP replied:
Good point. Fixed in b39a932ae1c220f5f188fde039d1a1d7228b4314

jketema commented:
> DCA reveals a few new results and a few lost results:
> 
>     * The lost `cpp/overrun-write` and `cpp/path-injection` results are because both queries use _any_ guard as a barrier, and we now recognize more barriers.
> 
>     * The `cpp/redundant-null-check-simple` results are new TPs 🎉

What about `cpp/user-controlled-bypass`?

MathiasVP commented:
> What about `cpp/user-controlled-bypass`?

We just discussed this offline: There aren't any query result differences. The change is in the number of `#select` results, but the number of query results remain the same.

jketema reviewed (approved):
LGTM
